# 코루틴(05) - 취소와 타입아웃(1)
### Job에 대해 취소
명시적인 Job에 대해 cancel 메서드를 호출해 취소할 수 있다.
```
import kotlinx.coroutines.*

suspend fun doOneTwoThree() = coroutineScope {
	val job1 = launch {
		println("launch1: ${Thread.currentThread().name}")
		delay(1000L)
		println("3!")
	}

	val job2 = launch {
		println("launch2: ${Thread.currentThread().name}")
		println("1!")
	}

	val job3 = launch {
		println("launch3: ${Thread.currentThread().name}")
		delay(500L)
		println("2!")
	}

	delay(800L)
	job1.cancel()
	job2.cancel()
	job3.cancel()
	println("4!")
}

fun main() = runBlocking {
	doOneTwoThree()
	println("runBlocking: ${Thread.currentThread().name}")
	println("5!")
}
```

### 취소 불가능한 Job
아래의 예제는 취소가 불가능한 Job이다.

launch(Dispatchers.Default)는 그 다음 코드 블록을 다른 스레드에서 수행을 시킬 것이다. 
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		var i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}

	delay(200L)
	job1.cancel()
	println("doCount Done!")
}

fun main() = runBlocking {
	doCount()
}

결과 ->
1
doCount Done!
2
3
4
5
6
7
8
9
10
```

두가지 부분이 신경이 쓰인다.
1. job1이 취소든 종료든 다 끝난 이후에 doCount Done!을 출력하고 싶다.
2. 취소가 되지 않았다.
먼저 취소든 종료든 다 끝난 이후에 doCountDone!을 출력하자.

### cancel과 join
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		var i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}

	delay(200L)
	job1.cancel()
	job1.join()
	println("doCount Done!")
}

fun main() = runBlocking {
	doCount()
}

결과 ->
1
2
3
4
5
6
7
8
9
10
doCount Done!
```

### cancelAndJoin
cancel을 하고 join을 하는 일은 자주 일어나는 일이기 때문에 한번에 하는 cancelAndJoin이 준비되어 있다.
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		var i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}

	delay(200L)
	job1.cancelAndJoin()
	println("doCount Done!")
}

fun main() = runBlocking {
	doCount()
}

결과 ->
1
2
3
4
5
6
7
8
9
10
doCount Done!
```

### cancel 가능한 코루틴
isActive를 호출하면 해당 코루틴이 여전히 활성화된지 확인 할 수 있다.
isActive를 루프에 추가해보자.
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		var i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10 && isActive) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}

	delay(200L)
	job1.cancelAndJoin()
	println("doCount Done!")
}

fun main() = runBlocking {
	doCount()
}

결과 ->
1
2
doCount Done!
```
isActive를 사용하면 cancel 가능한 코드가 된다.


# 코루틴(06) - finally를 같이 사용
launch에서 자원을 할당한 경우에는 어떻게 정리해야 할까요?

suspend 함수들은 JobCancellationException를 발생하기 때문에 표준 try catch finally로 대응할 수 있다.

```
import kotlinx.coroutines.*

suspend fun doOneTwoThree() = coroutineScope {
	val job1 = launch {
		try {
			println("launch1: ${Thread.currentThread().name}")
			delay(1000L)
			println("3!")
		} finally {
			println("job1 is finishing!")
			// 파일을 닫아주는 코드
		}
	}

	val job2 = launch {
		try {
			println("launch2: ${Thread.currentThread().name}")
			delay(1000L)
			println("1!")
		} finally {
			println("job2 is finishing!")
			// 소켓을 닫아주는 코드
		}
	}
	
	val job3 = launch {
		try {
			println("launch2: ${Thread.currentThread().name}")
			delay(1000L)
			println("2!")
		} finally {
			println("job3 is finishing!")
		}
	}

	delay(800L)
	job1.cancel()
	job2.cancel()
	job3.cancel()
	println("4!")
}

fun main() = runBlocking {
	doOneTwoThree()
	println("runBlocking: ${Thread.currentThread().name}")
	println("5!")
}

결과 ->
launch1: main @coroutine#2
launch2: main @coroutine#3
launch3: main @coroutine#4
4!
job1 is finishing!
job2 is finishing!
job3 is finishing!
runBlocking: main @coroutine#1
5!
```

try 문에서 cancel이 나버리면 finally가 호출된다.

### 취소 불가능한 블록
어떤 코드는 취소가 불가능해야 한다.
withContext(NonCancellable)을 이용하면 취소 불가능한 블록을 만들 수 있다.
```
import kotlinx.coroutines.*

suspend fun doOneTwoThree() = coroutineScope {
	val job1 = launch {
		withContext(NonCancellable) {
			println("launch1: ${Thread.currentThread().name}")
			delay(1000L)
			println("3!")
		}
		delay(1000L)
		println("job1: end")
	}

	val job2 = launch {
		withContext(NonCacellable) {
			println("launch2: ${Thread.currentThread().name}")
			delay(1000L)
			println("1!")
		} 
		delay(1000L)
		println("job2: end")
	}
	
	val job3 = launch {
		withContext(NonCacellable) {
			println("launch2: ${Thread.currentThread().name}")
			delay(1000L)
			println("2!")
		} 
		delay(1000L)
		println("job3: end")
	}

	delay(800L)
	job1.cancel()
	job2.cancel()
	job3.cancel()
	println("4!")
}

fun main() = runBlocking {
	doOneTwoThree()
	println("runBlocking: ${Thread.currentThread().name}")
	println("5!")
}

결과 ->
launch1: main @coroutine#2
launch2: main @coroutine#3
launch3: main @coroutine#4
4!
3!
1!
2!
runBlocking: main @coroutine#1
5!
```

이런 취소 불가능한 블록을 만드는 이유는 
“어떤 자원은 무조건 해제 해야된다.” 이런 코드가 있다면 이런 블록을 만들어 사용할 수 있다.

취소 불가능한 코드를 finally 절에 사용할 수도 있다.

### 타임 아웃
일정 시간이 끝난 후에 종료하고 싶다면 withTimeout을 이용할 수 있다.
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		val i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10 && isActive) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}
}

fun main() = runBlocking {
	withTimeout(500L) {
		doCount()
	}
}

결과 ->
1
2
3
4
Exception in thread "main"
kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 500ms...
```
이것을 try-catch로 검사하기는 매우 번거로운 일이다..

### withTimeoutOrNull
withTimeoutOrNull을 이용해 타임 아웃할 때 null을 반환하게 할 수 있다.
```
import kotlinx.coroutines.*

suspend fun doCount() = coroutineScope {
	val job1 = launch(Dispatchers.Default) {
		val i = 1
		var nextTime = System.currentTimeMillis() + 100L

		while(i <= 10 && isActive) {
			val currentTime = System.currentTimeMillis()
			if (currentTime >= nextTime) {
				println(i)
				nextTime = currentTime + 100L
				i++
			}
		}
	}
}

fun main() = runBlocking {
	val result = withTimeoutOrNull(500L) {
		doCount()
		true
	} ?: false
	println(result)
}

결과 ->
1
2
3
4
false
```
